src/payment/coupon/data/sagas.test.js

```diff
@@ -0,0 +1,246 @@
+import { runSaga } from 'redux-saga';
+
+import { configureApiService } from './service';
+import saga, { handleAddCoupon, handleRemoveCoupon } from './sagas';
+import { addCoupon, addCouponBegin, addCouponSuccess, addCouponFailure, removeCouponBegin, removeCouponSuccess, removeCoupon, removeCouponFailure } from './actions';
+import { PERCENTAGE_BENEFIT } from './constants';
+
+describe('saga tests', () => {
+  describe('main saga', () => {
+    it('should setup its sub sagas correctly', () => {
+      const gen = saga();
+
+      expect(gen.next()).toMatchSnapshot();
+      expect(gen.next()).toMatchSnapshot();
+      expect(gen.next()).toEqual({ done: true, value: undefined });
+    });
+  });
+
+  describe('handleAddCoupon', () => {
+    it('should handle success', async () => {
+      const responseData = {
+        data: {
+          voucher: {
+            id: 12345,
+            code: 'DEMO25',
+            benefit: {
+              type: PERCENTAGE_BENEFIT,
+              value: 25,
+            },
+          },
+        },
+      };
+
+      const apiClientPost = jest.fn(() =>
+        new Promise((resolve) => {
+          resolve(responseData);
+        }));
+
+      configureApiService(
+        {
+          ECOMMERCE_BASE_URL: 'http://localhost',
+        },
+        {
+          post: apiClientPost,
+        },
+      );
+
+      const dispatched = [];
+      await runSaga(
+        { dispatch: action => dispatched.push(action) },
+        handleAddCoupon,
+        addCoupon('DEMO25'),
+      ).toPromise();
+
+      expect(dispatched).toEqual([
+        addCouponBegin(),
+        addCouponSuccess(12345, 'DEMO25', {
+          type: PERCENTAGE_BENEFIT,
+          value: 25,
+        }),
+      ]);
+    });
+
+    it('should handle an API error', async () => {
+      const errorResponse = {
+        response: {
+          data: {
+            error_code: 'uhoh',
+          },
+        },
+      };
+
+      const apiClientPost = jest.fn(() =>
+        new Promise((resolve, reject) => {
+          reject(errorResponse);
+        }));
+
+      configureApiService(
+        {
+          ECOMMERCE_BASE_URL: 'http://localhost',
+        },
+        {
+          post: apiClientPost,
+        },
+      );
+
+      const dispatched = [];
+      await runSaga(
+        {
+          dispatch: action => dispatched.push(action),
+        },
+        handleAddCoupon,
+        addCoupon('DEMO25'),
+      ).toPromise();
+
+      expect(dispatched).toEqual([addCouponBegin(), addCouponFailure('uhoh')]);
+    });
+
+    it('should handle an unexpected error', async () => {
+      const errorResponse = 'oh snap!'; // We don't know how to handle this
+
+      const apiClientPost = jest.fn(() =>
```

Comment:
Might be nice to put the mock api client code together in a file or at the top of this one as a common object test can use.

Reply:
Another option is to make an object of the possible responses up at the top:

```
const responses = {
  success: { ....},
  errorNoKittens: {...},
  errorTooManyKittens: {....}
};```

It does move the responses away from your tests, but depending on the tests, maybe the details don't matter much and just the name is enough to show what you're testing.  (And it serves as a tidy documentation of what you're expecting to see.)

Reply:
Part of the problem is that if you put the mock setup at the top outside the tests, you need to reset them manually each time in a `beforeEach`.  Seemed more direct to just put the mock creation right there in the test, since it differs for each anyway.  

That said, I'm pulling the response objects out; that feels nice/makes the actual test code look shorter/less daunting.  

I struggled for a while to make this readable and settled on being verbose... thank you both for pushing to make it better. :D

Reply:
Part of the problem is that if you put the mock setup at the top outside the tests, you need to reset them manually each time in a `beforeEach`.  Seemed more direct to just put the mock creation right there in the test, since it differs for each anyway.  

That said, I'm pulling the response objects out; that feels nice/makes the actual test code look shorter/less daunting.  

I struggled for a while to make this readable and settled on being verbose... thank you both for pushing to make it better. :D